Add CUDA support for the OFI MTL

wckzhang opened the pull request:
This patch series touches the CUDA, opal datatype, PML CM, MTL datatype, and OFI MTL code.

The important things added here are:
Public function to allocate CUDA memory for datapacking of heterogenous buffers and device-to-device copy.
Pointer attribute detection in the CM PML layer.
Selecting libfabric providers with FI_HMEM and FI_MR_HMEM capabilities.
Only compiling the OFI MTL if libfabric version >= 1.9 if CUDA support is requested.
Memory registration for CUDA buffers.

bosilca commented on opal/datatype/opal_datatype_cuda.c:
```diff
@@ -111,6 +111,59 @@ bool  opal_cuda_check_one_buf(char *buf, opal_convertor_t *convertor )
     return ( ftable.gpu_is_gpu_buffer(buf, convertor));
 }
 
+/*
```
The OPAL datatype packing and unpacking is too low level to provide support for memory allocation. I understand this seems to be a popular place to put CUDA functions, but I would suggest these functions move to opal/mca/common/cuda.

bosilca commented on opal/datatype/opal_datatype_cuda.h:
```diff
@@ -18,12 +18,16 @@ struct opal_common_cuda_function_table {
     int (*gpu_cu_memcpy)(void*, const void*, size_t);
     int (*gpu_memmove)(void*, void*, size_t);
+    int (*gpu_malloc)(void*, size_t);
+    int (*gpu_free)(void*);
```
Technically I think it makes much more sense to move the entire CUDA structure outside of the datatype (especially now that it start having additional capabilities).

wckzhang replied:
Hmm, thanks for the feedback, I'll see what I can do about it.

gpaulsen commented:
@rajachan, you mentioned on Tuesday that AWS is doing some additional evaluation on this PR, is that correct?

hppritcha reviewed (approved):
I approve but agree with George's comments.

wckzhang commented:
Updated, there was a minor bug in the malloc code which I fixed (Didn't use a double pointer), and just basically moved all the opal_datatype_cuda.c and opal_datatype_cuda.h files into the corresponding common_cuda files.

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/970849af86693490a4531136ac9e4be8

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e13faff41eb79d6f2230ba9dfca596e4

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8423dc40cc266d16a37d32334aa7c6b9

wckzhang commented:
Looks like there's something I'm missing with compilation. Didn't hit this error during personal testing, not sure what's different in the CI setup compared to the manual build.

```
make[2]: *** No rule to make target 'mca/common/cuda/libmca_common_cuda.la', needed by 'libopen-pal.la'.  Stop.
```

gpaulsen commented:
@wckzhang did you figure it out?  I was going to try a local ppc64le build if you're stuck.

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/26475e9bedc1d8ff0754ee6cbe4ac127

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/cc495858babc4b613fdf123abf6e89c3

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ba79b0db3c1bbe0726e4cd6e546ce0bf

wckzhang commented:
I fixed the first error have another bug with cuda.h now it looks like

wckzhang commented:
oh yeah I'm pretty sure I know what's up, I only tested with cuda compiled, will add checks only if compiled with cuda.

wckzhang commented:
Bit of a dumb solution but it compiled properly both with and without cuda.

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/eea625079b1fa0922b8b1763d51206f0

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3f258c11df63d27640631c009cc9bae8

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e13e080e4383fc2410059a3d23bae826

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/04c803512823728b34dc32f84746452f

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/75d0a227795ea5b0d1fa4ba342c09735

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ff4dbb87c599b309a70414807dccd5c0

wckzhang commented:
Hmm, I don't see what the error is this time in the output, are the files truncated?

jjhursey commented:
It looks like it hit a 20-minute timeout on the make stage. Let me adjust that then I'll ask for a retest

jjhursey commented:
bot:ibm:retest

hppritcha commented:
Unfortunately this PR breaks a bunch of IBM tests which involve derived, non-contiguous datatypes when using the OFI mtl.

hppritcha commented on ompi/mca/pml/cm/pml_cm_sendreq.h:
```diff
@@ -242,6 +242,14 @@ do {                                                                    \
+        MCA_PML_CM_SWITCH_CUDA_CONVERTOR_OFF(flags, datatype, count);   \
+        (req_send)->req_base.req_convertor.flags |= flags;              \
+        /* Sets CONVERTOR_CUDA flag if CUDA buffer */                   \
+        opal_convertor_prepare_for_send(                                \
```
this broke use of OFI MTL in when not supporting CUDA.

wckzhang replied:
Can you describe what the exact test matrix was, ie. OMPI compiled w/ CUDA using a provider that supports/doesn't support FI_HMEM, using host buffers?

wckzhang replied:
Is this the same issue as your previous comment? This segment seems to only affect contiguous memory it looks like, does removing this section of code fix your non-contiguous use case?

bosilca replied:
Please read my comment on #8906 about this code.
